File: src/dotnet-interactive-vscode/src/vscode/diagnostics.ts
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

import * as vscode from 'vscode';

let diagnosticCollectionMap: Map<string, vscode.DiagnosticCollection> = new Map();

export function getDiagnosticCollection(cellUri: vscode.Uri): vscode.DiagnosticCollection {
    const key = cellUri.toString();
    let collection = diagnosticCollectionMap.get(key);
    if (!collection) {
        collection = vscode.languages.createDiagnosticCollection();
        diagnosticCollectionMap.set(key, collection);
    }

    collection.clear();
    return collection;
}

File: src/Microsoft.DotNet.Interactive.CSharpProject/Diagnostics.cs
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using Microsoft.CodeAnalysis;

namespace Microsoft.DotNet.Interactive.CSharpProject;

public class Diagnostics : ReadOnlyCollection<SerializableDiagnostic>, IRunResultFeature
{
    public Diagnostics(IList<SerializableDiagnostic> list) : base(list.Where(d => d.Severity is not DiagnosticSeverity.Hidden).ToList())
    {
    }

    public string Name => nameof(Diagnostics);

    public void Apply(FeatureContainer result)
    {
        result.AddProperty("diagnostics", this.Sort());
    }
}